File: lab09/lab09example01/scripts/lab09example01_node.py
#!/usr/bin/env python

import rospy
import random
from sensor_msgs.msg import JointState
from kdl_kine.kdl_kine_solver import  robot_kinematic
from kdl_kine.urdf import *
import numpy as np

def show_jacobian():

    rospy.init_node('kuka_dynamics_node')

    rate = rospy.Rate(10)
    # Initiate the base link(where the kinematic chain starts) and the end-effector link(where it ends).
    # Please look at the urdf file of the respective robot arm for the names.
    iiwa_kine = robot_kinematic('iiwa_link_0', 'iiwa_link_ee')

    q = PyKDL.JntArray(7)
    qdot = PyKDL.JntArray(7)

    while not rospy.is_shutdown():

        
        #Because joint_state_publisher does not publish velocity, we manually generate random joint positions and velocities as input to B, C, g matrices.
        for i in range(7):
            q[i] = random.uniform(-1, 1)
            qdot[i] = random.uniform(-0.5, 0.5)

        print('The current joint position: ')
        print(q)
        print('The current joint velocity: ')
        print(qdot)
        print('The current robot pose: ')
        print(iiwa_kine.forward_kinematics(q))
        print('B(q): ')
        #PyKDL has a unique representation for the matrix B (only in Python, doesn't apply to C++), so we have to convert the representation to a matrix manually.
        B = iiwa_kine.getB(q)

        Bmat = np.zeros((7, 7))
        for i in range(7):
            for j in range(7):
                Bmat[i, j] = B[i, j]
        print(Bmat)

        print('C(q, qdot): ')
        print(iiwa_kine.getC(q, qdot))

        print('g(q): ')
        print(iiwa_kine.getG(q))

        rate.sleep()



if __name__ == '__main__':
    show_jacobian()

File: development/lab09/lab09example01/src/lab09example01_node.cpp
#include "ros/ros.h"
#include <tf2_eigen/tf2_eigen.h>
#include <sensor_msgs/JointState.h>
#include "kdl_kine/kdl_kine_solver.h"
#include <Eigen/Dense>

int main(int argc, char **argv)
{

    ros::init(argc, argv, "kuka_dynamics_node");
    ros::NodeHandle nh;

    robot_kinematic iiwa_kine;
    ros::Rate loop_rate(10);

    //Initiate the base link (where the kinematic chain starts) and the end-effector link (where it ends). Please look at the
    // urdf file of the respective robot arm for the names.
    std::string base_link_name = "iiwa_link_0";
    std::string ee_link_name = "iiwa_link_ee";

    iiwa_kine.init(base_link_name, ee_link_name);

    KDL::JntArray q, qdot;
    q.resize(7);
    qdot.resize(7);


    while (ros::ok())
    {
       //Because joint_state_publisher does not publish velocity, we manually generate random joint positions and velocities as input to B, C, g matrices.
        MatrixXd q_and_qdot = MatrixXd::Random(7, 2);
        for (int i = 0; i < 7; i++)
        {
            q.data(i) = q_and_qdot(i, 0);
            qdot.data(i) = q_and_qdot(i, 1);
        }
        
        std::cout << "The current joint position: " << std::endl;
        std::cout << q.data << std::endl;

        std::cout << "The current joint velocity: " << std::endl;
        std::cout << qdot.data << std::endl;

        std::cout << "The current pose: " << std::endl;
        std::cout << iiwa_kine.KDLfkine(q) << std::endl;
        
        std::cout << "B(q): " << std::endl;
        std::cout << iiwa_kine.getB(q) << std::endl;
        std::cout << "C(q, qdot): " << std::endl;
        std::cout << iiwa_kine.getC(q, qdot) << std::endl;
        std::cout << "g(q): " << std::endl;
        std::cout << iiwa_kine.getG(q) << std::endl;

        ros::spinOnce();
        loop_rate.sleep();
    }



    return 0;
}
